rustls/src/msgs/message.rs

```diff
@@ -180,10 +176,66 @@ impl OpaqueMessage {
     const MAX_PAYLOAD: u16 = 16_384 + 2048;
 
     /// Content type, version and size.
-    const HEADER_SIZE: u16 = 1 + 2 + 2;
+    const HEADER_SIZE: usize = 1 + 2 + 2;
 
     /// Maximum on-the-wire message size.
-    pub const MAX_WIRE_SIZE: usize = (Self::MAX_PAYLOAD + Self::HEADER_SIZE) as usize;
+    pub const MAX_WIRE_SIZE: usize = Self::MAX_PAYLOAD as usize + Self::HEADER_SIZE;
+}
+
+#[derive(Clone, Debug)]
+pub struct PrefixedPayload(Vec<u8>);
+
+impl PrefixedPayload {
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut prefixed_payload = Vec::with_capacity(OpaqueMessage::HEADER_SIZE + capacity);
+        prefixed_payload.resize(OpaqueMessage::HEADER_SIZE, 0);
+        Self(prefixed_payload)
+    }
+}
+
+impl AsRef<[u8]> for PrefixedPayload {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[OpaqueMessage::HEADER_SIZE..]
+    }
+}
+
+impl AsMut<[u8]> for PrefixedPayload {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0[OpaqueMessage::HEADER_SIZE..]
+    }
+}
+
+impl<'a> Extend<&'a u8> for PrefixedPayload {
+    fn extend<T: IntoIterator<Item = &'a u8>>(&mut self, iter: T) {
+        self.0.extend(iter)
+    }
+}
+
+impl Deref for PrefixedPayload {
```

Comment:
Seems surprising that `Deref` yields different data from `AsRef`. Why do we need this `impl`?

Reply:
Yes, it's a bit awkward. The reason is in most places I want to be able to access "after prefix buffer" (with `AsRef`), but in some cases, I have to access the container (to extend the vec for example).
I like the solution of @ctz which requires less impls, I will take some inspiration from it.

Reply:
Not entirely sure what solution you're referring to. I think if this potential for confusion exists (prefix-less vs full backing store access) I think we should avoid `AsRef` and `Deref` equally and instead use explicitly named methods.

Reply:
`AsMut` and `Extend` have to be implemented because `PrefixedPayload` is passed to `seal_in_place_append_tag`.
I implemented `Deref` because we need to get the underlying container for those 3 operations used in multiple places by the `encrypt` methods:
```rs
payload.extend_from_slice(&nonce.as_ref()[4..]);
msg.payload.copy_to_vec(&mut payload);
msg.typ.encode(&mut payload);
```
It would be great to not expose the vector at all if possible. For the first operation, we can simply implement `extend_from_slice` on `PrefixedPayload`. For the second one, we can implement an `extend_from_chunks` method on `PrefixedPayload` which takes an `OutboundChunks` in. It avoids accessing the vector and I like the syntax better (as it resembles `copy_from_slice`). The last one is a bit more problematic. I would suggest modifying the `Codec` trait from:
```rs
pub trait Codec<'a>: Debug + Sized {
    fn encode(&self, bytes: &mut Vec<u8>);
    fn read(_: &mut Reader<'a>) -> Result<Self, InvalidMessage>;
    fn get_encoding(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        self.encode(&mut bytes);
        bytes
    }
    fn read_bytes(bytes: &'a [u8]) -> Result<Self, InvalidMessage> {
        let mut reader = Reader::init(bytes);
        Self::read(&mut reader)
    }
}
```
to:
```rs
trait Codec2<'a>: Debug + Sized {
    fn encode(&self, bytes: &mut Vec<u8>) {
        bytes.extend_from_slice(self.as_slice())
    }
    fn get_encoding(&self) -> Vec<u8> {
        self.as_slice().to_vec()
    }
    fn read_bytes(bytes: &'a [u8]) -> Result<Self, InvalidMessage> {
        let mut reader = Reader::init(bytes);
        Self::read(&mut reader)
    }
    fn as_slice(&self) -> &[u8];
    fn read(_: &mut Reader<'a>) -> Result<Self, InvalidMessage>;
}
```
In the end the 3 operations I listed would be rewritten:
```rs
payload.extend_from_slice(&nonce.as_ref()[4..]);
payload.extend_from_chunks(&msg.payload);
payload.extend_from_slice(msg.typ.as_slice());
```

Reply:
I just realized `as_slice` wouldn't work as it would return a reference to temporary data.

Reply:
Or instead of modifying `Codec` we could modify the macro `enum_builder` to generate a `get_slice` method (in addition of `get_u8`/`get_u16`).
So the code would look like:
```rs
payload.extend_from_slice(&msg.type.get_slice());
```
Another way would be to modify `Codec` to allow anything that implements `Extend<u8>` inside `encode` (not just `Vec<u8>`). But this is more involved and I don't know if the compiler is as efficient with `extend` as it can be with `extend_from_slice` or `push` and other vector-specific methods.
Or we could just add a `get_container` method on `PrefixedPayload` with a warning saying it contains the prefix as well as the body.

Reply:
I implemented the modified `enum_builder` on this branch: https://github.com/Wonshtrum/rustls/tree/message-types-rename.

Reply:
> we could modify the macro enum_builder to generate a get_slice method (in addition of get_u8/get_u16).

For what it's worth, I like that idea best of the set provided above. :+1:

Reply:
> I implemented `Deref` because we need to get the underlying container for those 3 operations used in multiple places by the `encrypt` methods:

Note to clarify, I'm not saying we should not allow access to the underlying container, merely that we shouldn't use the `Deref` trait implementation for it if it might be ambiguous if `Deref` yields the "full" container or just the payload part.

Reply:
Applied the `get_slice` approach from 5cfd6c49c92618a38cdc0004d40f43297da83d4e, but with the enum builder fn renamed to `get_array` to address Ctz's commit level feedback. I collapsed it into the commit introducing the relevant types instead of landing the deref version and updating it subsequently.